spray-http/src/main/scala/spray/http/HttpRequestRedactor.scala

```diff
@@ -0,0 +1,18 @@
+package spray.http
+
+/** Removes sensitive data from a request prior to outputting it e.g. in logs */
+trait HttpRequestRedactor {
+  def redact(req: HttpRequest): HttpRequest
+}
+
+object HttpRequestRedactor {
+  private var redactors = Set.empty[HttpRequestRedactor]
+
+  def add(r: HttpRequestRedactor): Unit = this.synchronized {
```

Comment:
Is there a use case for making the lis of redactors configurable / mutable at runtime?

Reply:
> request objects are created in many places in Spray where passing the implicit value may be difficult without extensive refactoring.

:+1:  I perfectly understand and ...

> Hence the idea of a global setting where new redactors can be added but none can be removed, so that this concern that is orthogonal to application code is handled consistently and automatically throughout the entire code base once the redactors have been set up.

... is a nice sweet spot to work around that.

What about making the configuration "impossible" to change once set? At least with checks at runtime as with something as suggested above?

```scala
def setRedactors(redactors: Set[HttpRequestRedactor]): Unit = if(this.redactors.isEmpty) {
 this.redactors = Some(redactors)
} else log.warn("redactors already configured...") // Or throw an exception (Philip Walder forgive me fot this :P)
```

Reply:
I'd like to allow different parts of an application to be able to register redactors related to their particular concerns.

If we enforce a "set only once" policy, then we require some top-level entity in the application to have the knowledge of all the various bits that need to be redacted in all the modules it aggregates, with the risk of missing some of them.

The current approach, based on a `Set` to which new elements can be added but none can be removed allows every module to contribute to redacting requests.